src/core/geometry/qgsinternalgeometryengine.cpp

```diff
@@ -1645,3 +1645,985 @@ QgsGeometry QgsInternalGeometryEngine::orientedMinimumBoundingBox( double &area,
+        yOutBuffer[1] = pY + side * amplitude * cosAngle;
+        amplitude = uniformDist( mt ) * ( maximumAmplitude - minimumAmplitude ) + minimumAmplitude;
+        xOutBuffer[2] = pX - side * amplitude * sinAngle;
+        bufferIndex = 2;
+        yOutBuffer[2] = pY - side * amplitude * cosAngle;
+        distanceToNextPointFromStartOfSegment += wavelength / 4;
+      }
+      else if ( bufferIndex == 2 )
+      {
+        xOutBuffer[3] = pX - side * amplitude * sinAngle;
+        yOutBuffer[3] = pY - side * amplitude * cosAngle;
+
+        if ( isFirstPart )
+        {
+          xOutBuffer[2] = ( xOutBuffer[2] + xOutBuffer[3] ) * 0.5;
+          yOutBuffer[2] = ( yOutBuffer[2] + yOutBuffer[3] ) * 0.5;
+          isFirstPart = false;
+        }
+
+        // flush curve
+        std::unique_ptr< QgsLineString > bezier( QgsLineString::fromBezierCurve( QgsPoint( xOutBuffer[0], yOutBuffer[0] ),
+            QgsPoint( xOutBuffer[1], yOutBuffer[1] ),
+            QgsPoint( xOutBuffer[2], yOutBuffer[2] ),
+            QgsPoint( xOutBuffer[3], yOutBuffer[3] ),
+            segments ) );
+        out->append( bezier.get() );
+
+        // shuffle buffer alone
+        xOutBuffer[0] = xOutBuffer[3];
+        yOutBuffer[0] = yOutBuffer[3];
+        bufferIndex = 1;
+        side = -side;
+
+        wavelength = uniformDist( mt ) * ( maximumWavelength - minimumWavelength ) + minimumWavelength;
+
+        distanceToNextPointFromStartOfSegment += wavelength / 4;
+      }
+    }
+    totalCoveredDistance += segmentLength;
+
+    prevX = thisX;
+    prevY = thisY;
+    distanceToNextPointFromStartOfSegment -= segmentLength;
+  }
+
+  const double midX = prevX - remainingDistance / 2 * std::sin( segmentAngleRadians );
+  const double midY = prevY - remainingDistance / 2 * std::cos( segmentAngleRadians );
+  const double pX = midX + side * amplitude * std::sin( segmentAngleRadians + M_PI_2 );
+  const double pY = midY + side * amplitude * std::cos( segmentAngleRadians + M_PI_2 );
+
+  std::unique_ptr< QgsLineString > bezier( QgsLineString::fromBezierCurve( out->endPoint(),
+      QgsPoint( ( out->endPoint().x() + pX ) * 0.5, ( out->endPoint().y() + pY ) * 0.5 ),
+      QgsPoint( ( prevX + pX ) * 0.5, ( prevY + pY ) * 0.5 ),
+      QgsPoint( prevX, prevY ),
+      segments ) );
+  out->append( bezier.get() );
+
+  return out;
+}
+
+std::unique_ptr< QgsAbstractGeometry > roundWavesPrivate( const QgsAbstractGeometry *geom, double wavelength, double amplitude, bool strictWavelength )
+{
+  std::unique_ptr< QgsAbstractGeometry > segmentizedCopy;
+  if ( QgsWkbTypes::isCurvedType( geom->wkbType() ) )
+  {
+    segmentizedCopy.reset( geom->segmentize() );
+    geom = segmentizedCopy.get();
+  }
+
+  if ( QgsWkbTypes::geometryType( geom->wkbType() ) == QgsWkbTypes::LineGeometry )
+  {
+    return roundWavesAlongLine( static_cast< const QgsLineString * >( geom ), wavelength, amplitude, strictWavelength );
+  }
+  else
+  {
+    // polygon
```

Comment:
Where are we guarding ourselves from the abstract geometry pointer being a point geometry? Should we check for point type and return an empty abstract geometry here (and in the two other triangle/square private functions?

Reply:
It's done in the public functions -- e.g. at https://github.com/qgis/QGIS/pull/45583/files/d9e3659879c8ab168302755a68bb13f561b2b68f#diff-17cd6748c845e0043f19426bb01e791a11d7a0310044693d2fcb4a08d5c5c4c9R2560
So here we are guaranteed to have line or polygon types (the public functions also handle the multipart types, so again we are guaranteed here to have a single-part, linear type). There's unit tests covering all this too for safety.